master/internal/provisioner/aws_spot_collections.go

```diff
@@ -0,0 +1,188 @@
+	delete(c.keyMap, s)
+	return c
+}
+
+// deleteIntersection delete any items that exist in both set and set2 from set.
+func (c *setOfSpotRequests) deleteIntersection(set2 setOfSpotRequests) *setOfSpotRequests {
+	for _, req := range set2.iter() {
+		c.delete(req)
+	}
+	return c
+}
+
+// contains checks if this spotRequest is in the set
+func (c *setOfSpotRequests) contains(s *spotRequest) bool {
+	_, ok := c.keyMap[s.SpotRequestId]
+	return ok
+}
+
+// containsId checks if a spotRequest with this id is in the set
+func (c *setOfSpotRequests) containsId(s string) bool {
+	_, ok := c.keyMap[s]
+	return ok
+}
+
+// copy creates a shallow copy of the set
+func (c *setOfSpotRequests) copy() setOfSpotRequests {
+	set := newSetOfSpotRequests()
+	for _, req := range c.iter() {
+		set.add(req)
+	}
+	return set
+}
+
+// asList return the spotRequests in the set as a slice
+func (c *setOfSpotRequests) asList() []*spotRequest {
+	list := make([]*spotRequest, 0, len(c.keyMap))
+	for _, sr := range c.keyMap {
+		list = append(list, sr)
+	}
+	return list
+}
+
+// asListInChronologicalOrder returns the spotRequests in the set as a slice,
+// sorted in chronological order
+func (c *setOfSpotRequests) asListInChronologicalOrder() []*spotRequest {
+	l := c.asList()
+	sort.SliceStable(l, func(i, j int) bool {
+		return l[i].CreationTime.Before(l[j].CreationTime)
+	})
+	return l
+}
+
+// idsAsList returns the spotRequest ids as a slice of strings
+func (c *setOfSpotRequests) idsAsList() []string {
+	list := make([]string, 0, len(c.keyMap))
+	for reqId := range c.keyMap {
+		list = append(list, reqId)
+	}
+	return list
+}
+
+// idsAsListOfPointers returns the spotRequest ids as a slice of string pointers
+func (c *setOfSpotRequests) idsAsListOfPointers() []*string {
+	list := make([]*string, 0, len(c.keyMap))
+	for reqId := range c.keyMap {
+		b := strings.Builder{}
+		b.WriteString(reqId)
+		k := b.String()
+		list = append(list, &k)
+	}
+	return list
+}
+
+// iter returns the underlying map to make iterating over setOfSpotRequests as clean as
+// iterating over a map. e.g.:
+// for reqId, req := range set.iter() { ... }
+func (c *setOfSpotRequests) iter() map[string]*spotRequest {
+	return c.keyMap
+}
+
+// numReqs returns the number of spotRequests in the set
+func (c *setOfSpotRequests) numReqs() int {
+	return len(c.keyMap)
+}
+
+// string returns a string representation of the set, which is the list of ids separated by commas
+func (c *setOfSpotRequests) string() string {
+	l := c.idsAsList()
+	return strings.Join(l, ",")
+}
+
+// newSetOfSpotRequests creates a new, empty setOfSpotRequests
+func newSetOfSpotRequests() setOfSpotRequests {
+	return setOfSpotRequests{
+		keyMap: make(map[string]*spotRequest),
+	}
+}
+
+// setOfStrings is a light wrapper around a map to make it look like a set. Primarily
+// exists to hide golang boilerplate behind descriptively named functions. Elements
+// in the set are strings.
+type setOfStrings struct {
```

Comment:
I feel like it's pretty standard Golang practice to wrap a `map` in a `Set` when you need set semantics. `map` is too low level to allow for good code here - the Golang boilerplate hides the business logic behind too many lines of code focused on how exactly you are using the `map` to perform that business logic.

Reply:
Why is it a standard Golang practice? It doesn't sound very inefficient to implement a set for all of the types since there are no generics. We should be consistent in such uses across the codebase. And writing a set doesn't really save code lines. :) It just allows semantic names for set operations.

Reply:
BTW, we generally follow this [doc](https://docs.google.com/document/d/1V0m1V7pU1nInAVFe5Vg2Hcj7hxcMGrINRjg1scTKJPQ/edit#), which we have settled on following effective go to use `map[...]bool` as a set. We can discuss it in #infra-ag if you think it's not right.